Add `strict-dynamic` CSP tests, based on SecurityPolicyViolationEvent.

mikispag opened the pull request:
Adding `strict-dynamic` CSP tests, not based on `checkReport.sub.js`, but on `SecurityPolicyViolationEvent`.

The tests are covering pretty tricky use cases, and so far partially fail in Chromium stable and beta, and fully pass in Chromium Canary. Firefox is implementing `strict-dynamic` in [this Bugzilla issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1299483).

mikispag committed:
Add `strict-dynamic` CSP tests, based on SecurityPolicyViolationEvent.

sideshowbarker commented:
paging @hillbrad @mikewest @metromoxie for review

mikispag committed:
Add a test case checking that a script injected via `appendChild` from a script matching SHA256 hash is allowed with `strict-dynamic`.

mikispag committed:
Rephrase violation reports check message.

mikispag committed:
Add a positive test case for the double policy (`strict-dynamic` & nonce+whitelist).

mikewest commented:
I'm OOO today, but I'll take a look at this tomorrow.

On Wednesday, 19 October 2016, Michael[tm] Smith <notifications@github.com>
wrote:

> paging @hillbrad <https://github.com/hillbrad> @mikewest
> <https://github.com/mikewest> @metromoxie <https://github.com/metromoxie>
> for review
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/4020#issuecomment-254757535>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAF2XFTI0T9Ag0aFNGzAaztZ989xrGaks5q1d57gaJpZM4Kauf9>
> .
>


-- 
-mike

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_and_unsafe_eval_eval.html:
```diff
@@ -0,0 +1,39 @@
+        async_test(function(t) {
+            setTimeout(t.step_func_done(function() {
+                assert_no_violations();
+            }), 1000);
```
A second is a long time to wait for automated tests. Since there's no way to prove that the event _never ever_ fires, I'd suggest that it's enough to add something like `document.addEventListener('securitypolicyviolation', t.unreached_func("No violation event should fire."));` to the first test.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_and_unsafe_eval_new_function.html:
```diff
@@ -0,0 +1,39 @@
+            setTimeout(t.step_func_done(function() {
+                assert_no_violations();
+            }), 1000);
+        }, 'No CSP violation report has been fired.');
```
Ditto.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_discard_whitelist.html:
```diff
@@ -0,0 +1,39 @@
+                assert_violations({
+                    'script-src': 1
+                });
+            }), 1000);
```
Ditto.

I'd prefer for this to be structured as a single test. That is:

```
<script nonce="dummy">
  async_test(function(t) {
    document.addEventListener("securitypolicyviolation", t.step_func_done(function (e) {
      assert_false(simpleSourcedScriptRan);
      assert_equals(e.violatedDirective, "script-src");
    }));
  });
</script>
```

That captures both the tests you're running in a way that doesn't involve timeouts.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_different_nonce.html:
```diff
@@ -0,0 +1,56 @@
+                    'script-src': 1 // Only one violation, coming from the second policy.
+                });
+                assert_original_policies(["script-src 'nonce-dummy' 'nonce-dummy2'"]);
+            }), 1000);
```
As you're probably noticing by now, I don't like this pattern. :) I'll stop commenting on it, as I imagine it shows up below as well, but I'd prefer to get rid of it when possible.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_different_nonce.html:
```diff
@@ -0,0 +1,56 @@
+            e.src = 'simpleSourcedScript.js';
+            e.onload = t.unreached_func('OnLoad should not be triggered.');
+            document.body.appendChild(e);
+            t.done();
```
Calling `done` here means that you're not giving `onload` time to execute. I think you could be more precise by hooking the `onerror` event instead:

```
e.onload = t.unreached_func();
e.onload = t.step_func_done();
document.body.appendChild(e).
```

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_different_nonce.html:
```diff
@@ -0,0 +1,56 @@
+            e.setAttribute('nonce', 'dummy2');
+            e.src = 'simpleSourcedScript.js';
+            e.onload = t.step_func_done(function() {
+                assert_true(simpleSourcedScriptRan);
```
Hrm. You have two tests looking at the same global state. Would it be possible to have the script alter its environment in some non-shared way? For instance: `document.currentScript.setAttribute('executed', '1');` followed by `assert_equals(e.getAttribute('executed'), '1');`.

mikewest commented on content-security-policy/support/checkSecurityPolicyViolations.sub.js:
```diff
@@ -0,0 +1,33 @@
+var cspViolationReports = [];
```
Why does this file need substitution (that is, why name it `.sub.js`)?

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_honor_whitelist.html:
```diff
@@ -0,0 +1,55 @@
+    <script nonce='dummy'>
+        async_test(function(t) {
+            var e = document.createElement('script');
+            e.src = 'http://example.org/script.js';
```
Rather than `example.org` (which exists, and would cause network requests if the test failed), please use something like `{{location[scheme]}}://{{domains[www2]}}:{{ports[http][0]}}` to point to one of WPT's aliases. (Note: you'll need to rename the file to `.sub.html` for that substitution to work).

mikewest reviewed (changes requested):
Thanks for putting this together! I've left a few comments based on a review of the first few files. Skimming the rest, the general `setTimeout(..., 1000)` pattern is just too expensive for this many tests. I think you can get rid of it in practically every case, and doing so will make it more likely that we can actually bring theses tests into an automated CI system. As-is, these will just take too long to run.

mikispag committed:
Implement reviewer-suggested changes.

mikispag commented:
PTAL.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_discard_whitelist.html:
```diff
@@ -0,0 +1,27 @@
+        async_test(function(t) {
+            document.addEventListener('securitypolicyviolation', t.step_func_done(function(e) {
+                assert_equals(document.getElementById('whitelistedScript').getAttribute('executed'), null);
+                assert_equals(e.violatedDirective.split(' ')[0], 'script-src');
```
Why do you need the `split()`? Are you seeing multiple directives in a report? If so, that seems like a bug.

mikispag replied:
The violatedDirective field is like: script-src 'self' http: ... etc. Since Firefox expands 'self' with the full origin, i.e. http://origin.tld:port, I am just checking the actual directive, so 'script-src'. It's not my idea, it is exactly like that in existing CSP tests.

mikispag replied:
I mean, I know I could use substitution, but I would have to do that conditionally, just for FF. I would rather just check the directive alone, as other tests do. WDYT?

mikewest replied:
Hrm. I think this is a bug in Chrome.

Look at `e.effectiveDirective` instead, which should just be a directive name. Is Firefox doing the same thing as Chrome? We might just need to live with the wierdness if so. :(

mikispag replied:
IIRC Firefox had issues with effectiveDirective in the past (sometimes it was not populated) and it was more reliable to rely on the content of violatedDirective before the first space. But it looks like FF behaves like Chrome for an explicit script-src, so I am switching to that.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_different_nonce.html:
```diff
@@ -0,0 +1,55 @@
+                assert_equals(document.getElementById('nonced-appendChild').getAttribute('executed'), '1');
+            });
+            e.onerror = t.unreached_func('Error should not be triggered.');
+            document.body.appendChild(e);
```
This test isn't testing anything. It exits right after this line, and neither `onload` nor `onerror` fire in time.

Rather than having a single handler, why not just have multiple? That is, the first test could look like:

```
var e = document.createElement('script');
e.id = ...
e.src = ...
e.onload = ...

document.addEventListener('securitypolicyviolation', t.step_func(function (e) {
  if (e.target != e)
    return;
  ...
  t.done();
}));

document.body.appendChild(e);
```

You could do the same in the second test as well, and then each handler will only finish the test once its element is responsible for a violation, or doesn't execute (in which case you'd change `t.done()` to `assert_unreached(...)`).

mikispag replied:
I am not sure I understand, I already have:

```
document.addEventListener('securitypolicyviolation', t.unreached_func('No CSP violation report has fired.'));
```

in the second test, and it is the only listener at that point. I want to make sure that does not trigger, and that's all. How is your solution different?

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_honor_whitelist.sub.html:
```diff
@@ -0,0 +1,59 @@
+        async_test(function(t) {
+            var handler = function(e) {
+                document.removeEventListener('securitypolicyviolation', handler);
+                t.unreached_func('No CSP violation report has fired.');
```
`assert_unreached`. `t.unreached_func()` will return a function. You'd need to write `(t.unreached_func())()` for that to work.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_report_only.html:
```diff
@@ -0,0 +1,39 @@
+            var handler = function(e) {
+                assert_equals(e.violatedDirective.split(' ')[0], 'script-src');
+                // Check that every violation comes from the Report-Only policy.
+                assert_equals(e.originalPolicy.trim(), "script-src 'none'");
```
Ditto with the call to `trim()`. Are you seeing extra whitespace?

mikispag replied:
Yes, I do.

In `script-src-strict_dynamic_double_policy_honor_whitelist.sub.html`, I get:

`Uncaught Error: assert_equals: expected "script-src 'self' 'nonce-dummy'" but got " script-src 'self' 'nonce-dummy'"`

I don't know if it is due to the double policy, I am serving:

```
Content-Security-Policy: script-src 'strict-dynamic' 'nonce-dummy'
Content-Security-Policy: script-src 'self' 'nonce-dummy'
```

Is it a Chromium (Canary) bug? Shall I assert without .trim()?

mikewest replied:
Seems like a bug to me. Firefox is doing it right, I assume?

mikispag replied:
I am sorry, it was my fault. Please ignore this.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_meta_tag.html:
```diff
@@ -0,0 +1,94 @@
+        e.setAttribute('executed', '0');
+        e.textContent = "document.currentScript.setAttribute('executed', '1');";
+        e.onerror = t.unreached_func('Error should not be triggered.');
+        var observer = new MutationObserver(function(mutations) {
```
This is a bit complicated. Did you consider just putting a test step function onto the global that the script could call? Like:

```
var e = document.createScript('script');
...
window.appendChildViaTextContent = t.step_func_done();
e.textContent = "appendChildViaTextContent();"
```

That leaves one global dangling around for each test, but seems more straightforward.

mikispag replied:
I knew there was a simpler way than using mutation observers :) .

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_parser_inserted.html:
```diff
@@ -0,0 +1,170 @@
+        }, 'Parser-inserted script via `document.write` without a correct nonce is not allowed with `strict-dynamic`.');
+    </script>
+
+    <!-- This test must be synchronous because we have to wait until the old event handler is removed. -->
```
See comments above about another way of approaching this, as I don't think these tests are testing anything.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_parser_inserted_correct_nonce.html:
```diff
@@ -0,0 +1,86 @@
+    <script nonce='dummy'>
+        async_test(function(t) {
+            document.writeln('<scr' + 'ipt defer async nonce="dummy" id="documentWriteln-defer-async" src="simpleSourcedScript.js"></scr' + 'ipt>');
+            assert_not_equals(document.getElementById('documentWriteln-defer-async'), null);
```
These are only checking that the element exists, right? Shouldn't you be checking at some point that the relevant element has a value set on it? Perhaps another approach would be to have a script that posts a message to itself: something like:

```
// amazing-script-name.js
window.postMessage(document.currentScript.id, "*");
```

and

```
// amazing test
async_text(function (t) {
  window.addEventListener("message", t.step_func(function (e) {
    if (e.data == "element-id-from-below")
      t.done();
  }));
  document.write("<script id='element-id-from-below' ...></scr" + "ipt>");
}, "Testy test.");
```

WDYT?

mikewest reviewed (changes requested):
This is looking really good! I have a few suggestions for even more improvements.

mikispag committed:
Get rid of all synchronous tests by adding multiple handlers to the CSP violation event, make simpleSourcedScript.js send a message with its ID.

mikispag commented:
PTAL. I might not have addressed all your concerns, but since I significantly changed things following your advice, it is better to get comments on newer code.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_parser_inserted.html:
```diff
@@ -0,0 +1,193 @@
+    <script nonce='dummy'>
+        var insertAdjacentHTMLScriptRan = false;
+        async_test(function(t) {
+            window.addEventListener('securitypolicyviolation', t.step_func_done(function(violation) {
```
I think these should all be `step_func()` rather than `step_func_done()`. The latter will complete the test when _any_ `securitypolicyviolation` event is fired (even if you return early without doing any assertions). Switching to `step_func()` and calling `t.done()` manually after your assertions should mean that you're actually testing each block individually.

mikewest reviewed (approved):
LGTM, with one change I'd like you to make before landing. Thanks for going a few rounds on this! I think the tests have ended up much improved!

mikispag committed:
Fix SecurityPolicyViolation event handlers, no longer checking e.target.

mikispag commented:
PTAL. Happy that this unveiled bugs in Chromium!

sideshowbarker commented:
[FYI the reason the stability checker is failing on the https://github.com/w3c/web-platform-tests/pull/4020/commits/e7675d000db06e0427a0af6bfbd8f76030823b74 change is just because there’s a bug/regression in the stability-checker source on master that I inadvertently introduced. The fix for that is waiting in https://github.com/w3c/web-platform-tests/pull/4097 for review. Once that has landed (sometime in the next 6 hours or so) it should work as expected for subsequent commits you make on this branch.]

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_discard_whitelist.html:
```diff
@@ -20,8 +20,9 @@
-            window.addEventListener('securitypolicyviolation', t.step_func_done(function(e) {
+            window.addEventListener('securitypolicyviolation', t.step_func(function(e) {
                 assert_equals(e.violatedDirective.split(' ')[0], 'script-src');
+                t.done();
```
Nit: This change doesn't seem necessary. You're not exiting the handler early in any case; there's only one event, and when it fires, you're done. `step_func_done` seems like the right thing to use here.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_different_nonce.html:
```diff
@@ -40,17 +40,16 @@
                 }
             }));
 
             window.addEventListener('securitypolicyviolation', t.step_func(function(violation) {
```
Isn't this going to fire when the handler in the first block fires? You're attaching two listeners that expect opposite things. I don't understand how that will work.

mikewest commented on content-security-policy/script-src/script-src-strict_dynamic_double_policy_honor_whitelist.sub.html:
```diff
@@ -17,18 +17,18 @@
                 }
             }));
-            window.addEventListener('securitypolicyviolation', t.step_func(function(violation) {
-                if (violation.target.id !== 'whitelisted-appendChild') {
```
Note that for URL-based violations, you can check `e.blockedURI` for these kinds of checks.

mikewest reviewed (changes requested):
This change seems like a regression. :( Changing from `async_test` to `test` seems like it's going to make your tests not actually test anything. Likewise, the handler logic seems very strange.

mikispag committed:
Revert "Fix SecurityPolicyViolation event handlers, no longer checking e.target."

This reverts commit e7675d000db06e0427a0af6bfbd8f76030823b74.

mikispag committed:
Use e.blockedURI when it is not possible to use e.target.

mikispag commented:
Sorry for the extra round, IIUC your comments, it should be good now. Note that script-src-strict_dynamic_double_policy_honor_whitelist.sub.html fails in Chromium Canary because of https://codereview.chromium.org/2451673002.

mikispag committed:
Add: `strict-dynamic` does not drop whitelists in `img-src`.

mikispag commented:
Please LMK if I should add a test for 'strict-dynamic' in 'default-src', and, if yes, how.

mikewest commented:
Hi, I'm sorry; I lost track of this, then I was OOO, and then I continued to forget about it. :( Taking a look now.

mikewest reviewed (approved):
LGTM, thank you for going through so many iterations. :)
